policy/compiler.go

```diff
@@ -0,0 +1,199 @@
+}
+
+// Compile generates a single CEL AST from a collection of policy expressions associated with a CEL environment.
+func Compile(env *cel.Env, p *Policy) (*cel.Ast, *cel.Issues) {
+	c := &compiler{
+		env:  env,
+		info: p.SourceInfo(),
+		src:  p.Source(),
+	}
+	errs := common.NewErrors(c.src)
+	iss := cel.NewIssuesWithSourceInfo(errs, c.info)
+	rule, ruleIss := c.compileRule(p.Rule(), c.env, iss)
+	iss = iss.Append(ruleIss)
+	if iss.Err() != nil {
+		return nil, iss
+	}
+	ruleRoot, _ := env.Compile("true")
+	opt := cel.NewStaticOptimizer(&ruleComposer{rule: rule})
+	ruleExprAST, iss := opt.Optimize(env, ruleRoot)
+	return ruleExprAST, iss.Append(iss)
+}
+
+func (c *compiler) compileRule(r *Rule, ruleEnv *cel.Env, iss *cel.Issues) (*compiledRule, *cel.Issues) {
+	var err error
+	compiledVars := make([]*compiledVariable, len(r.Variables()))
+	for i, v := range r.Variables() {
+		exprSrc := c.relSource(v.Expression())
+		varAST, exprIss := ruleEnv.CompileSource(exprSrc)
+		if exprIss.Err() == nil {
+			ruleEnv, err = ruleEnv.Extend(cel.Variable(fmt.Sprintf("variables.%s", v.Name().Value), varAST.OutputType()))
+			if err != nil {
+				iss.ReportErrorAtID(v.Expression().ID, "invalid variable declaration")
+			}
+			compiledVars[i] = &compiledVariable{
+				name: v.name.Value,
+				expr: varAST,
+			}
+		}
+		iss = iss.Append(exprIss)
+	}
+	compiledMatches := []*compiledMatch{}
+	for _, m := range r.Matches() {
+		condSrc := c.relSource(m.Condition())
+		condAST, condIss := ruleEnv.CompileSource(condSrc)
+		iss = iss.Append(condIss)
+		if m.HasOutput() && m.HasRule() {
+			iss.ReportErrorAtID(m.Condition().ID, "either output or rule may be set but not both")
+			continue
+		}
+		if m.HasOutput() {
+			outSrc := c.relSource(m.Output())
+			outAST, outIss := ruleEnv.CompileSource(outSrc)
+			iss = iss.Append(outIss)
+			compiledMatches = append(compiledMatches, &compiledMatch{
+				cond:   condAST,
+				output: outAST,
+			})
+			continue
+		}
+		if m.HasRule() {
+			nestedRule, ruleIss := c.compileRule(m.Rule(), ruleEnv, iss)
+			iss = iss.Append(ruleIss)
+			compiledMatches = append(compiledMatches, &compiledMatch{
+				cond:       condAST,
+				nestedRule: nestedRule,
+			})
+		}
+	}
+	return &compiledRule{
+		variables: compiledVars,
+		matches:   compiledMatches,
+	}, iss
+}
+
+func (c *compiler) relSource(pstr ValueString) *RelativeSource {
+	line := 0
+	col := 1
+	if offset, found := c.info.GetOffsetRange(pstr.ID); found {
+		if loc, found := c.src.OffsetLocation(offset.Start); found {
+			line = loc.Line()
+			col = loc.Column()
+		}
+	}
+	return c.src.Relative(pstr.Value, line, col)
+}
+
+type ruleComposer struct {
+	rule *compiledRule
+}
+
+// Optimize implements an AST optimizer for CEL which composes an expression graph into a single
+// expression value.
+func (opt *ruleComposer) Optimize(ctx *cel.OptimizerContext, a *ast.AST) *ast.AST {
+	ruleExpr, _ := optimizeRule(ctx, opt.rule)
+	ctx.UpdateExpr(a.Expr(), ruleExpr)
```

Comment:
Oh, I'll add a comment. The optimize step here is weird because it starts off with `true` and replaces the expression completely with the policy expression fragments

Reply:
I'm primarily looking to see if the optimized policies can actually be unparsed. Relooking at the UpdateExpr code, I feel like it actually would not populate the binds correctly for the macro source, but maybe I'm wrong.

Reply:
I've added round-tripping on the generated expressions. There was a small bug where the last macro was not being added into the list due to how I was updating the root expression.